### PR TITLE
devtools: Fix infinite load in devtools package

### PR DESCRIPTION
**Pull request by jikim-msft**
#### Description

https://dev.azure.com/fluidframework/internal/_workitems/edit/5176/

This PR updates the logic in `devtools-core` package to correctly handle visualization for `unknown` objects.

### CODE REVIEW

File: packages/tools/devtools/devtools-core/src/data-visualization/DataVisualization.ts [hunk 1]
```diff
@@ -212,11 +212,13 @@ export class DataVisualizerGraph
 		const result: Record<string, RootHandleNode> = {};
 		await Promise.all(
 			rootDataEntries.map(async ([key, value]) => {
-				const fluidObjectId = await this.registerVisualizerForHandle(value.handle);
-				result[key] =
-					fluidObjectId === undefined
-						? unknownObjectNode
-						: createHandleNode(fluidObjectId);
+				if (value.handle !== undefined) {
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/data-visualization/DataVisualization.ts:215]: In the case where `value.handle` is undefined, we should do 2 things (in my opinion):
1. Log an error to the console
2. Set `result[key]` to `unknownObjectNode`

This way we give ample warning to the user that they've configured something wrong.

File: packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts [hunk 2]
```diff
@@ -205,4 +206,67 @@ describe("DataVisualizerGraph unit tests", () => {
...
+			runtime,
+			SharedCell.getFactory().attributes,
+		);
+		sharedCell.set("Hello world");
+
+		const unknownObject = {};
+
+		const visualizer = new DataVisualizerGraph(
+			{
+				counter: sharedCounter,
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts:231]: Once we're generating an unknown object node for the invalid data case (see my other comment), we can probably simplify this test quite a bit by _only_ specifying the invalid data here.

Comment (jikim-msft) [reply packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts]: I changed the unit test to only contain the `containerData` which is `unknown` to the visualizer. Also added an additional test case which passes an empty `containerData` to the `visualizer` for sanity check.

File: packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts [hunk 3]
```diff
@@ -205,4 +206,29 @@ describe("DataVisualizerGraph unit tests", () => {
...
+		const visualizer = new DataVisualizerGraph(
+			{
+				unknownObject: unknownObject as IFluidLoadable,
+			},
+			defaultVisualizers,
+			defaultEditors,
+		);
+
+		const childUnknownObject = await visualizer.render("unknown-object");
+		expect(childUnknownObject).to.deep.equal(undefined);
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts:222]: This should be `unknownObjectNode`, right? Not `undefined`?

File: packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts [hunk 4]
```diff
@@ -205,4 +206,29 @@ describe("DataVisualizerGraph unit tests", () => {
...
+
+		const visualizer = new DataVisualizerGraph(
+			{
+				unknownObject: unknownObject as IFluidLoadable,
+			},
+			defaultVisualizers,
+			defaultEditors,
+		);
+
+		const childUnknownObject = await visualizer.render("unknown-object");
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts:221]:
```suggestion
		const childUnknownObject = await visualizer.render("unknownObject");
```

Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts:221]: This should call `visualizer.renderRoot` (or whatever that method is called). The one that gives us all our root handles back. Then we can verify that it contains a single entry `unknownObject: unknownObjectNode`.

File: packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts [hunk 5]
```diff
@@ -205,4 +206,29 @@ describe("DataVisualizerGraph unit tests", () => {
...
+		expect(childUnknownObject).to.deep.equal(undefined);
+	});
+
+	it("Empty Container Data", async () => {
+		// Pass in the empty containerData to the visualizer.
+		const emptyRecord: Record<string, IFluidLoadable> = {};
+
+		const visualizer = new DataVisualizerGraph(emptyRecord, defaultVisualizers, defaultEditors);
+
+		const childEmptyRecord = await visualizer.render("empty-containerData");
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts:231]: This only verifies that we can't find something to render for key "empty-containerData". We want to verify that rendering the root handles of the visualizer returns an empty record when the containerData we gave it was empty.

File: packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts (hunk 4 above)
Comment (jikim-msft) [reply packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts]: Yes! I should have made the `childUnknownObject` to be `unknownObjectNode` as the `renderRootHandles()` would mark is as `unknownObjectNode`. 

I initially made a mistake of calling `render()` instead of `renderRootHandles()` to retrieve all the root handles. I have updated the logic.

File: packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts (hunk 5 above)
Comment (jikim-msft) [reply packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts]: I made the following changes for the `emptyRecord` unit test:
- Call `renderRootHandles()`
- Compare the output with the empty record of `expectedChildEmptyRecord`

One thing that I am wondering is whether or not setting the type as `Record<string, RootHandleNode>` is correct.

`renderRootHandles()` method first initializes the empty Record as `const result: Record<string, RootHandleNode> = {};` and would return this right away as the `rootDataEntries` is empty (and this is why I used `RootHandleNode` as the expected type). 

But since `RootHandleNode` can either be `FluidHandleNode | UnknownObjectNode`, I am unsure if I should set more of less-generic type.

File: packages/tools/devtools/devtools-core/src/data-visualization/DataVisualization.ts [hunk 6]
```diff
@@ -212,11 +212,16 @@ export class DataVisualizerGraph
...
+					const fluidObjectId = await this.registerVisualizerForHandle(value.handle);
+					result[key] =
+						fluidObjectId === undefined
+							? unknownObjectNode
+							: createHandleNode(fluidObjectId);
+				} else {
+					console.error("containerData contains unknown data object!");
```
Comment (Josmithr) [inline packages/tools/devtools/devtools-core/src/data-visualization/DataVisualization.ts:222]: Nit: I think we can be a bit more specific here. Maybe something like \`Container data includes a non-Fluid object under key ${key}. Cannot visualize.\`

**Review by Josmithr — APPROVED** [review]
Left one more nitpicky comment regarding the error message we log in the failure case. Otherwise 🚀

File: packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts (hunk 5 above)
Comment (Josmithr) [reply packages/tools/devtools/devtools-core/src/test/DataVisualizerGraph.test.ts]: Honestly, you could probably just pass in `{}` (and ignore any linter errors). I don't think you'll have to specify a type in that case.
